File: java/patterns/src/main/java/org/prototypepatterns/behavioral/observer/ObserverPattern.java
package org.prototypepatterns.behavioral.observer;

import java.util.ArrayList;
import java.util.List;
import java.util.function.IntConsumer;

public final class ObserverPattern {
    private ObserverPattern() {
    }

    public static final class Subject {
        private final List<IntConsumer> observers = new ArrayList<>();

        public void subscribe(IntConsumer observer) {
            observers.add(observer);
        }

        public void setState(int value) {
            observers.forEach(observer -> observer.accept(value));
        }
    }

    public static List<Integer> demo() {
        List<Integer> events = new ArrayList<>();
        Subject subject = new Subject();
        subject.subscribe(events::add);
        subject.setState(7);
        return events;
    }
}

File: csharp/patterns/src/PrototypePatterns/Behavioral/Observer/ObserverPattern.cs
namespace PrototypePatterns.Behavioral.Observer;

public sealed class Subject
{
    private event Action<int>? Changed;

    public void Subscribe(Action<int> observer)
    {
        Changed += observer;
    }

    public void SetState(int value)
    {
        Changed?.Invoke(value);
    }
}

public static class ObserverPattern
{
    public static IReadOnlyList<int> Demo()
    {
        var events = new List<int>();
        var subject = new Subject();
        subject.Subscribe(events.Add);
        subject.SetState(7);
        return events;
    }
}
